docs: document documentation process (#7891)
near/nearcore: docs/practices/docs.md

# Documentation

> This chapter describes nearcore approach to develope documentation. In general,

Reviewer: ```suggestion
This chapter describes nearcore's approach to documentation. In general,
```

> there are three primary types of documentation to keep in mind:

Reviewer: ```suggestion
there are three primary types of documentations to keep in mind:
```

* **Protocol specification** ([source](https://github.com/near/NEPs),
  [rendered](https://nomicon.io)) is a formal description of the NEAR protocol.
  It can be used to implement alternative NEAR clients.

> * **User Docs** ([rendered](https://docs.near.org)) explain how to use the near

Reviewer: ```suggestion
* **User docs** ([rendered](https://docs.near.org)) explain how to use the near
```

  network. User docs are also split into documentation for validators (that is,
  how to run your own near node) and documentation for smart contract
  developers.

> * **Internal Development Docs** ([rendered](https://near.github.io/nearcore/),

Reviewer: ```suggestion
* **Internal development docs** ([rendered](https://near.github.io/nearcore/),
```

  [source](https://github.com/near/nearcore/tree/master/docs)) is the book you

>   are reading right now! The target audience here are developers of nearcore.

Reviewer: nit: I think it is easy to misread "developers of nearcore" and deduce something like developers building on top of near.  I saw this happen in a couple of conversations.  I would rephrase it like something below

```suggestion
  are reading right now! The target audience here are nearcore contributors.
```

> Here, we focus on the last sort -- internal docs.

Reviewer: I think we can remove this line.  We say above that this book is about internal docs.

```suggestion
```

[here](https://rust-lang.github.io/mdBook/), but the basics are very simple.

To add a new page:

1. Add an `.md` file somewhere in the
   [`./docs`](https://github.com/near/nearcore/tree/master/docs) folder.
2. Add an entry to
   [`SUMMARY.md`](https://github.com/near/nearcore/blob/master/docs/SUMMARY.md)
   file.
3. Submit a PR (again, we promise to merge it without much ceremony).

The doc itself is vanilla markdown.

To render documentation locally:

```console
# Install mdBook
$ cargo install mdbook
$ mdbook serve --open ./docs
```

This will convert `.md` file from the docs folder to `.html`, open a browser,
and start a file watcher to rebuild and reload on change.

Note that GitHub's default rendering mostly just works as well, so you don't
need special preview when reviewing pull requests to docs.

The book deployed via this GitHub Action:
[.github/workflows/book.yml](https://github.com/near/nearcore/blob/master/.github/workflows/book.yml).
It just runs mdBook and then deploys the result to
https://near.github.io/nearcore/.

For internal docs, you often want to have pretty pictures. We don't currently
have a recommended workflow, but here are some tips:

> * Don't add binary media files to Git. Rather, upload images as comments to this

Reviewer: nit: might be worth it to add a sentence to explain why we do not linke binary files.

> * Don't add binary media files to Git. Rather, upload images as comments to this
>   super-secret issue [#7821](https://github.com/near/nearcore/issues/7821), and
>   then link to the images as

Author: ```suggestion
* Don't add binary media files to Git to avoid inflating repository size. 
  Rather, upload images as comments to this super-secret issue 
  [#7821](https://github.com/near/nearcore/issues/7821), and then link to 
  the images as
```
